opensquirrel/decomposer/aba_decomposer.py

```diff
@@ -0,0 +1,170 @@
+
+    _gate_list: list[Callable[..., BlochSphereRotation]] = [Rx, Ry, Rz]
+
+    def __init__(self) -> None:
+        self.index_a = self._gate_list.index(self.ra)
+        self.index_b = self._gate_list.index(self.rb)
+
+    def get_decomposition_angles(self, alpha: float, axis: AxisLike) -> tuple[float, float, float]:
+        """
+        Gives the angles used in the A-B-A decomposition of the Bloch sphere rotation
+        characterized by a rotation around `axis` of angle `alpha`.
+
+        Parameters:
+            alpha: angle of the Bloch sphere rotation
+            axis: _normalized_ axis of the Bloch sphere rotation
+
+        Returns:
+            A triple (theta1, theta2, theta3) corresponding to the decomposition of the
+            arbitrary Bloch sphere rotation into U = Ra(theta3) Rb(theta2) Ra(theta1)
+
+        """
+        axis = Axis(axis)
+
+        if not (-math.pi + ATOL < alpha <= math.pi + ATOL):
+            raise ValueError("Angle needs to be normalized")
+
+        if abs(alpha - math.pi) < ATOL:
+            # alpha == pi, math.tan(alpha / 2) is not defined.
+
+            p: float
+            if abs(axis[self.index_a]) < ATOL:
+                theta2 = math.pi
+                p = 0
+                m = 2 * math.acos(axis[self.index_b])
+
+            else:
+                p = math.pi
+                theta2 = 2 * math.acos(axis[self.index_a])
+
+                if abs(axis[self.index_a] - 1) < ATOL or abs(axis[self.index_a] + 1) < ATOL:
+                    m = p  # This can be anything, but setting m = p means theta3 == 0, which is better for gate count.
+                else:
+                    m = 2 * math.acos(axis[self.index_b] / math.sqrt(1 - axis[self.index_a] ** 2))
+
+        else:
+            p = 2 * math.atan2(axis[self.index_a] * math.sin(alpha / 2), math.cos(alpha / 2))
+
+            acos_argument = math.cos(alpha / 2) * math.sqrt(1 + (axis[self.index_a] * math.tan(alpha / 2)) ** 2)
+
+            # This fixes float approximations like 1.0000000000002 which acos doesn't like.
+            acos_argument = max(min(acos_argument, 1.0), -1.0)
+
+            theta2 = 2 * math.acos(acos_argument)
+            theta2 = math.copysign(theta2, alpha)
+
+            if abs(math.sin(theta2 / 2)) < ATOL:
+                m = p  # This can be anything, but setting m = p means theta3 == 0, which is better for gate count.
+            else:
+                acos_argument = float(axis[self.index_b]) * math.sin(alpha / 2) / math.sin(theta2 / 2)
+
+                # This fixes float approximations like 1.0000000000002 which acos doesn't like.
+                acos_argument = max(min(acos_argument, 1.0), -1.0)
+
+                m = 2 * math.acos(acos_argument)
+
+        theta1 = (p + m) / 2
+
+        theta3 = p - theta1
+        return theta1, theta2, theta3
+
+    def decompose(self, g: Gate) -> list[Gate]:
+        if not isinstance(g, BlochSphereRotation):
+            # Only decomposer single-qubit gates.
```

Comment:
We are not checking if `g` is a single qubit gate, we are checking if it is a `BlochSphereRotation`. Either this comments should change, or the `get_qubit_operands` operands method should be used.

Reply:
I added a length checker. Placing a filter for gates that are not an instance of BlochSphereRotation is beneficial for MyPy. If we do not have an isinstance checker, MyPy complains that we are using methods that do not exist on the Gate object i.e gate.angle() and gate.axis()